File: js/intro/ejercicios-finales/ejercicio-final-02-RESUELTO.js
// 2. Añade un parámetro a la función para poder pasarle un listado de stopwords. 
//   slugify(texto, ['de','que','los'])
//   Ejemplos:
//   * ' que significan los sueños' --> '/significan-suenos'
//   * 'qué hacer después de programar' --> '/hacer-despues-programar'


const mapping = {
  '-': [' ','/','_',',',':',';'],
  'a': ['à','á','ä','â'],
  'e': ['è','é','ë','ê'],
  'i': ['ì','í','ï','î'],
  'o': ['ò','ó','ö','ô'],
  'u': ['ù','ú','ü','û'],
  'n': ['ñ'],
  'c': ['ç']
}

function slugify(str,stopwords) {
  let letters = str.toLowerCase();

  if(stopwords) {
    stopwords.forEach(stopword => {
        letters = letters.replace(` ${stopword} `, ' ')
    });
  }

  letters = letters.split('');

  const goodLetters = letters.map(letter => {
    for (const [key, value] of Object.entries(mapping)) {
      if(value.includes(letter)) {
        return key;
      }
    }
    return letter;
  })

  return goodLetters
    .join('')
    .replace(/-{2,}/g, '-')
    .replace(/^-|-$/g, '')
    .trim()
}

File: python/intro/ejercicios-finales/ejercicio-final-02-RESUELTO.py
# 2. Añade un parámetro a la función para poder pasarle un listado de stopwords. 
#   slugify(texto, ['de','que','los'])
#   Ejemplos:
#   * ' que significan los sueños' --> '/significan-suenos'
#   * 'qué hacer después de programar' --> '/hacer-despues-programar'
#
#  TIP: ignoraremos las stopwords si estan al principio o al final

mapping = {
  '-': [' ','/','_',',',':',';'],
  'a': ['à','á','ä','â'],
  'e': ['è','é','ë','ê'],
  'i': ['ì','í','ï','î'],
  'o': ['ò','ó','ö','ô'],
  'u': ['ù','ú','ü','û'],
  'n': ['ñ'],
  'c': ['ç']
}

def iterReplace(letter):
  for key, value in mapping.items():
    if letter in value:
      return key
  return letter

def slugify(texto, stopwords = None):
  letters = texto.lower().strip()

  if stopwords:
    for stopword in stopwords:
      letters = letters.replace(' ' + stopword + ' ', ' ')

  goodLetters = list(map(iterReplace, letters))
  return ''.join(goodLetters)

slugify('qué hacer después de programar', ['de', 'después'])